Added Decorations

rm-dr opened the pull request:
# Preview:
![image](https://user-images.githubusercontent.com/96270320/206881704-c6b6cc74-6939-4364-943f-e2d6c15c3284.gif)

![everything](https://user-images.githubusercontent.com/96270320/208325339-cebfc960-f3b8-411c-a641-22558e71ab08.gif)

```elixir
# Top gif
Output out.gif
Set FontSize 20
Set Width 600
Set Height 300
Set Padding 20

Set MarginFill image.jpg
Set MarginSize 30

Type "This is demo text"
Sleep 2s
```

```elixir
# Bottom gif
Output out.gif
Set FontSize 15
Set Width 600
Set Height 300
Set Padding 20

Set MarginFill "#79B8C3"
Set MarginSize 40

Set WindowBar Colorful
Set WindowBarSize 40

Set CornerRadius 10

Type "I have a bar, a margin, and rounded corners."
Sleep 2s
```

Added margins, round corners, and window bars. Addresses most of #173.

# Features:
- Adds settings `MarginSize`, `MarginFill`, `WindowBar`, `WindowBarSize`, `CornerRadius`
- Dimensions of the output are exactly `width` and `height`, terminal size is calculated
- Errors are raised when `width` or `height` are too small


# Other notes:
 - This does not gracefully exit when a margin image doesn't exist. I left that out intentionally, should be fixed with #179.
- I did not add an example to the main `README.md` since the examples there are linked from `stuff.charm.sh`, and I do not have access to that. It might be a good idea to remove the examples in that file completely, since they're already in `examples/*`. There's no need to have *two* copies of every example tape!
- It may be good to add a warning about gif palettes to the docs. Margin images with many colors turn out pretty ugly.

adriens commented:
OMG :heavy_heart_exclamation: What a beauty :clap:

maaslalani commented:
This is awesome, @rm-dr! Thank you so much for this PR.

Some feedback on this feature and thinking about the language interface: the thing I think of when I see this is an image background, rather than a frame. When I think of a frame I moreso think of the border of the terminal (such as the green, red, yellow dots).

So, I think maybe we should rename this `Background`. I can see a future where we also also an option for terminal frames + backgrounds + margin. Something like:

```elixir
Set Frame MacOS
Set Background https://example.com/link-to-some-cool-background.png
Set Background #FFF # perhaps this can also take a color?
Set Margin 40 # the terminal frame will have 40px of background surrounding it
Set Padding 20 # the content will have 20px of the terminal background color surrounding it
```

adriens commented:
OH BOY, OH BOY ❣️ 🏋️ 🧑‍🎨

rm-dr commented:
Yes, that sounds like a good idea.

`ImageFrame` and `FrameSize` should be renamed, but `Background` is already used in multiple places to refer to the *terminal's* background (`Options.Video.BackgroundColor` and `Options.Theme.Background`, for example).

Adding an `Options.Video.Background` would make things a bit confusing.

We could use a different name internally, like `MarginBackground?`
That might be a clearer, or it might be more confusing than before.

rm-dr commented:
I renamed the setting keys and added static color support. It would be nice to add a format check for colors so a malformed hex string won't give a huge, hard-to-read ffmpeg error.

maaslalani commented:
This is really awesome @rm-dr, thank you _so_ much!

maaslalani commented on examples/settings/set-margin.tape:
```diff
@@ -0,0 +1,13 @@
+
+# You can also use an image:
+# Set MarginFill "path/to/image.png"
+Set MarginFill "#79B8C3"
+Set MarginSize 20
```
I _think_ we'll want to call these `Background` and `Margin` externally. Internally, in the code, we can keep these names.

```suggestion
Set Background "#79B8C3"
Set Margin 20
```

rm-dr replied:
Are you sure? I think that the `Background` setting key is a bit misleading, since backgrounds are usually "interior & padding" areas, not "margin" areas.

For example, this is how CSS works ([jsfiddle](https://jsfiddle.net/x3wcuabm/4/)).

Maybe run a poll on discord?

maaslalani replied:
Yeah, actually I think you are right, background might suggest the terminal background.

Another approach we could go for is calling it Border, what do you think?

```
Border 20px "#FFF"
```

or 

```
BorderSize 20px
BorderFill "#FFF"
```

maaslalani replied:
Perhaps `Margin` is the correct name, Border could be used in conjunction with the Margin I guess.

rm-dr replied:
Yeah, I think "Margin" makes the most sense
Border is terminal frame, padding is padding, etc

Just like the CSS box model we're used to!

maaslalani commented on video.go:
```diff
@@ -78,6 +85,148 @@ func DefaultVideoOptions() VideoOptions {
 }
 
+// BuildFFopts builds an ffmpeg command from a VideoOptions
+func buildFFopts(opts VideoOptions, targetFile string) []string {
```
Nice, thanks for making this better!

maaslalani commented:
Just as a heads up, we will likely need this error to be a `ParserError` so that it does the nice underlining for the color token.

<img width="675" alt="image" src="https://user-images.githubusercontent.com/42545625/208141771-e0032fce-5be2-42f1-b299-3636ed6cf69a.png">


Such as like this:

<img width="790" alt="image" src="https://user-images.githubusercontent.com/42545625/208142134-32d2e7e2-8060-414a-8b31-fc2e0f1f273e.png">

Let me know if you need any pointers on how to do that (or if you want me to implement this myself)

rm-dr commented:
I just added parse-time color checking, and may have found a minor bug:

Error highlighting doesn't account for quoted strings!
![image](https://user-images.githubusercontent.com/96270320/208162246-9f91a905-d6e0-4eab-8d1a-3a222dbd3aa2.png)

maaslalani commented:
> I just added parse-time color checking, and may have found a minor bug:
> 
> Error highlighting doesn't account for quotes surrounding strings! ![image](https://user-images.githubusercontent.com/96270320/208162246-9f91a905-d6e0-4eab-8d1a-3a222dbd3aa2.png)

Thanks! That looks awesome.

Yeah the off by one error definitely looks like a bug, should be an easy solve. It's unfortunate that `#` is the comment character which means the color must be quoted :/

maaslalani commented:
@rm-dr The new rounded corners and terminal frames looks so awesome! Thank you so much!

winder commented:
I gave this branch a try today. It worked as expected and the rendered recording looks great. Despite that there were errors reported:
```
Couldn't draw Bar, something is very wrong.
Couldn't draw CornerMask, something is very wrong.
```

I used the window configuration you have above:
```
Set MarginFill "#79B8C3"
Set MarginSize 40

Set WindowBar Colorful
Set WindowBarSize 40

Set CornerRadius 10
```

rm-dr commented:
Stupid bug. I put a `==` where a `!=` should be. Give me a second...

rm-dr commented:
That's odd. Yesterday's changes could not possibly affect margin scaling (or anything else, in fact).

You're right, vhs should force aspect ratio if the margin image does not match the video. I tried to reproduce this today, but couldn't get your results. Can you send your image/config/system info?

My test files are below.

<img src="https://user-images.githubusercontent.com/96270320/211597535-ac4c0c25-64ea-4510-a80c-b1c473390127.jpg" height="200">

```elixir
Output test.gif
Set FontSize 20
Set Width 600
Set Height 300
Set Padding 20

Set MarginFill test.jpg
Set MarginSize 50

Type "I have a margin!"
Sleep 2s
```

![test](https://user-images.githubusercontent.com/96270320/211598853-3307812e-43c0-4832-91fc-c55ad5b70c98.gif)

rm-dr commented:
I know this is a big addition to `vhs`, but it would be nice to have a merge ETA.

Is there anything else that needs to be done?

maaslalani commented:
Hey @rm-dr, thanks again so much for this awesome PR, I'll try and give it a first pass review today and try out all the features!

maaslalani commented:
Amazing work! This has been merged in via: https://github.com/charmbracelet/vhs/pull/274
